File: Backend/Service/Website/app/routers/bots.py
"""
Bot routes
"""
from fastapi import APIRouter, HTTPException, Depends, status
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import exists
from typing import List

from app.database import get_db
from app.models import Bot, Document, DocumentShare, bot_documents
from app.schemas.bot import BotCreate, BotUpdate, BotResponse
from app.dependencies import get_current_user
from app.models import User

router = APIRouter(prefix="/bots", tags=["bots"])

HELP_BOT_NAME = "บอทช่วยสอน"


def format_bot(bot: Bot) -> dict:
    """Format bot for response"""
    return {
        "id": bot.id,
        "name": bot.name,
        "prompt": bot.prompt,
        "description": bot.description,
        "model": bot.model,
        "avatarUrl": bot.avatarUrl,
        "enabled": bot.enabled if bot.enabled is not None else True,
        "ownerId": bot.ownerId,
        "createdAt": bot.createdAt,
        "updatedAt": bot.updatedAt,
        "documents": [
            {"id": doc.id, "displayName": doc.displayName}
            for doc in (bot.documents if bot.documents else [])
        ]
    }


@router.get("", response_model=List[BotResponse])
async def get_bots(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Get all bots owned by current user"""
    bots = (
        db.query(Bot)
        .options(joinedload(Bot.documents))
        .filter(Bot.ownerId == current_user.id)
        .order_by(Bot.createdAt.desc())
        .all()
    )
    
    # Check if user has help bot
    has_help_bot = any(b.name == HELP_BOT_NAME for b in bots)
    if not has_help_bot:
        help_bot = (
            db.query(Bot)
            .options(joinedload(Bot.documents))
            .filter(Bot.name == HELP_BOT_NAME)
            .first()
        )
        if help_bot:
            bots.append(help_bot)
    
    return [format_bot(bot) for bot in bots]


@router.post("", response_model=BotResponse, status_code=status.HTTP_201_CREATED)
async def create_bot(
    bot_data: BotCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Create a new bot"""
    try:
        if not bot_data.name or not bot_data.prompt:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="name and prompt are required"
            )
        
        # Validate document IDs if provided
        valid_document_ids = []
        if bot_data.documentIds:
            unique_ids = list(set([id for id in bot_data.documentIds if id]))
            if unique_ids:
                try:
                    # Check if user has access to documents (owner or shared)
                    accessible_docs = (
                        db.query(Document)
                        .filter(
                            Document.id.in_(unique_ids),
                            (
                                (Document.ownerId == current_user.id) |
                                (Document.shares.any(DocumentShare.userId == current_user.id))
                            )
                        )
                        .all()
                    )
                    valid_document_ids = [doc.id for doc in accessible_docs]
                except Exception as e:
                    print(f"⚠️  Error validating documents: {e}")
                    # Continue without documents if validation fails
                    valid_document_ids = []
        
        # Create bot
        from datetime import datetime
        now = datetime.now()
        bot = Bot(
            name=bot_data.name,
            prompt=bot_data.prompt,
            description=bot_data.description,
            model=bot_data.model,
            avatarUrl=bot_data.avatarUrl,
            ownerId=current_user.id,
            enabled=bot_data.enabled if bot_data.enabled is not None else True,
            createdAt=now,
            updatedAt=now
        )
        db.add(bot)
        db.flush()  # Get bot.id
        
        # Associate documents
        if valid_document_ids:
            try:
                for doc_id in valid_document_ids:
                    db.execute(
                        bot_documents.insert().values(botId=bot.id, documentId=doc_id)
                    )
            except Exception as e:
                print(f"⚠️  Error associating documents: {e}")
                # Rollback and continue without documents
                db.rollback()
                db.flush()
        
        db.commit()
        db.refresh(bot)
        
        # Load documents for response
        try:
            bot_with_docs = (
                db.query(Bot)
                .options(joinedload(Bot.documents))
                .filter(Bot.id == bot.id)
                .first()
            )
            if bot_with_docs:
                bot = bot_with_docs
        except Exception as e:
            print(f"⚠️  Error loading bot with documents: {e}")
            # Continue with bot without documents if loading fails
            pass
        
        # Return formatted bot (will be validated against BotResponse schema)
        return format_bot(bot)
    except HTTPException:
        raise
    except Exception as e:
        print(f"❌ Error creating bot: {e}")
        import traceback
        traceback.print_exc()
        db.rollback()
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating bot: {str(e)}"
        )


@router.patch("/{bot_id}", response_model=BotResponse)
async def update_bot(
    bot_id: int,
    bot_data: BotUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Update a bot"""
    bot = (
        db.query(Bot)
        .filter(Bot.id == bot_id, Bot.ownerId == current_user.id)
        .first()
    )
    
    if not bot:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bot not found"
        )
    
    if bot.name == HELP_BOT_NAME:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="ไม่สามารถแก้ไขบอทช่วยสอนได้"
        )
    
    # Update fields
    if bot_data.name is not None:
        bot.name = bot_data.name
    if bot_data.prompt is not None:
        bot.prompt = bot_data.prompt
    if bot_data.description is not None:
        bot.description = bot_data.description
    if bot_data.model is not None:
        bot.model = bot_data.model
    if bot_data.avatarUrl is not None:
        bot.avatarUrl = bot_data.avatarUrl
    if bot_data.enabled is not None:
        bot.enabled = bot_data.enabled
    
    # Update documents if provided
    if bot_data.documentIds is not None:
        # Delete existing associations
        db.execute(
            bot_documents.delete().where(bot_documents.c.botId == bot.id)
        )
        
        # Add new associations
        if bot_data.documentIds:
            unique_ids = list(set([id for id in bot_data.documentIds if id]))
            if unique_ids:
                accessible_docs = (
                    db.query(Document)
                    .filter(
                        Document.id.in_(unique_ids),
                        (
                            (Document.ownerId == current_user.id) |
                            (Document.shares.any(DocumentShare.userId == current_user.id))
                        )
                    )
                    .all()
                )
                valid_ids = [doc.id for doc in accessible_docs]
                for doc_id in valid_ids:
                    db.execute(
                        bot_documents.insert().values(botId=bot.id, documentId=doc_id)
                    )
    
    db.commit()
    db.refresh(bot)
    
    # Load documents for response
    bot = (
        db.query(Bot)
        .options(joinedload(Bot.documents))
        .filter(Bot.id == bot.id)
        .first()
    )
    
    return format_bot(bot)


@router.delete("/{bot_id}")
async def delete_bot(
    bot_id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    """Delete a bot"""
    bot = (
        db.query(Bot)
        .filter(Bot.id == bot_id, Bot.ownerId == current_user.id)
        .first()
    )
    
    if not bot:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Bot not found"
        )
    
    if bot.name == HELP_BOT_NAME:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="ไม่สามารถลบบอทช่วยสอนได้"
        )
    
    db.delete(bot)
    db.commit()
    
    return {"ok": True}

File: Backend/server/routes/bots.js
import express from "express";
import { prisma } from "../db.js";
import { authenticate } from "../lib/auth.js";
import { getRequestContext } from "../lib/requestContext.js";
import { logEvent } from "../lib/logging.js";
import { invalidateUserCaches } from "../lib/cache.js";
import { deleteBotWithCleanup } from "../services/uploadQueue.js";

export const botsRouter = express.Router();
const HELP_BOT_NAME = "บอทช่วยสอน";

const formatBot = (bot) => ({
  id: bot.id,
  name: bot.name,
  prompt: bot.prompt,
  description: bot.description,
  model: bot.model,
  avatarUrl: bot.avatarUrl,
  enabled: bot.enabled !== false,
  createdAt: bot.createdAt,
  updatedAt: bot.updatedAt,
  documents: (bot.documents || []).map((link) => link.document),
});

botsRouter.get("/", authenticate, async (req, res) => {
  const bots = await prisma.bot.findMany({
    where: { ownerId: req.user.id },
    orderBy: { createdAt: "desc" },
    include: {
      documents: {
        include: {
          document: { select: { id: true, displayName: true } },
        },
      },
    },
  });
  // ไม่โชว์บอทช่วยสอนในหน้ารายการ Bot (ใช้เบื้องหลังสำหรับ 3 ปุ่มบน homepage)
  const filtered = bots.filter((b) => b.name !== HELP_BOT_NAME);
  res.json(filtered.map(formatBot));
});

botsRouter.post("/", authenticate, async (req, res) => {
  const { name, prompt, description, model, avatarUrl, documentIds } = req.body ?? {};

  if (!name || !prompt) {
    res.status(400).json({ error: "name and prompt are required" });
    return;
  }

  const rawIds = Array.isArray(documentIds) ? documentIds : [];
  const uniqueIds = [...new Set(rawIds.filter(Boolean))];
  let validIds = uniqueIds;
  if (uniqueIds.length > 0) {
    const accessibleDocs = await prisma.document.findMany({
      where: {
        id: { in: uniqueIds },
        OR: [
          { ownerId: req.user.id },
          { shares: { some: { userId: req.user.id } } },
        ],
      },
      select: { id: true },
    });
    const accessibleIdSet = new Set(accessibleDocs.map((doc) => doc.id));
    validIds = uniqueIds.filter((id) => accessibleIdSet.has(id));
  }

  const bot = await prisma.bot.create({
    data: {
      name,
      prompt,
      description: description ?? null,
      model: model ?? null,
      avatarUrl: avatarUrl ?? null,
      ownerId: req.user.id,
      enabled: true,
      documents: validIds.length
        ? {
            create: validIds.map((id) => ({ documentId: id })),
          }
        : undefined,
    },
    include: {
      documents: {
        include: { document: { select: { id: true, displayName: true } } },
      },
    },
  });

  await logEvent({
    event: "bot.created",
    actorId: req.user.id,
    targetType: "bot",
    targetId: bot.id,
    meta: { name: bot.name, documentIds: validIds, documentCount: validIds.length, ...getRequestContext(req) },
  });

  res.status(201).json({
    id: bot.id,
    name: bot.name,
    prompt: bot.prompt,
    description: bot.description,
    model: bot.model,
    avatarUrl: bot.avatarUrl,
    enabled: true,
    createdAt: bot.createdAt,
    updatedAt: bot.updatedAt,
    documents: bot.documents.map((link) => link.document),
  });
  await invalidateUserCaches(req.user.id);
});

const HELP_DOC_DISPLAY_NAME = "คู่มือการใช้งาน";

// ไม่ต้องล็อกอิน — ให้ 3 ปุ่มบอทช่วยสอนบน homepage ใช้ได้เลย
botsRouter.get("/help-config", async (req, res) => {
  const helpDoc = await prisma.document.findFirst({
    where: { displayName: HELP_DOC_DISPLAY_NAME },
    select: { id: true },
  });
  if (!helpDoc) {
    return res.json({ botId: null, documentId: null, bot: null });
  }
  const candidates = await prisma.bot.findMany({
    where: {
      enabled: { not: false },
      documents: {
        some: { documentId: helpDoc.id },
      },
    },
    include: {
      documents: {
        include: {
          document: { select: { id: true, displayName: true } },
        },
      },
    },
  });
  const helpBot = candidates.find((b) => b.name === HELP_BOT_NAME) || candidates[0] || null;
  if (!helpBot) {
    return res.json({ botId: null, documentId: null, bot: null });
  }
  res.json({
    botId: helpBot.id,
    documentId: helpDoc.id,
    bot: formatBot(helpBot),
  });
});

botsRouter.get("/:id", authenticate, async (req, res) => {
  const bot = await prisma.bot.findFirst({
    where: {
      id: req.params.id,
      OR: [{ ownerId: req.user.id }, { name: HELP_BOT_NAME }],
    },
    include: {
      documents: {
        include: { document: { select: { id: true, displayName: true } } },
      },
    },
  });
  if (!bot) {
    res.status(404).json({ error: "Bot not found" });
    return;
  }
  res.json(formatBot(bot));
});

botsRouter.patch("/:id", authenticate, async (req, res) => {
  try {
    const { name, prompt, description, model, avatarUrl, documentIds, enabled } = req.body ?? {};

    const hasAny =
      name !== undefined || prompt !== undefined || documentIds !== undefined ||
      description !== undefined || model !== undefined || avatarUrl !== undefined || enabled !== undefined;
    if (!hasAny) {
      res.status(400).json({ error: "at least one of name, prompt, description, model, avatarUrl, documentIds, enabled is required" });
      return;
    }

    const bot = await prisma.bot.findFirst({
      where: { id: req.params.id, ownerId: req.user.id },
    });

    if (!bot) {
      res.status(404).json({ error: "Bot not found" });
      return;
    }

    if (bot.name === HELP_BOT_NAME) {
      res.status(403).json({ error: "ไม่สามารถแก้ไขบอทช่วยสอนได้" });
      return;
    }

    const rawIds = Array.isArray(documentIds) ? documentIds : null;
    const ids = rawIds ? [...new Set(rawIds.filter(Boolean))] : null;
    let validIds = ids;
    if (ids && ids.length > 0) {
      const accessibleDocs = await prisma.document.findMany({
        where: {
          id: { in: ids },
          OR: [
            { ownerId: req.user.id },
            { shares: { some: { userId: req.user.id } } },
          ],
        },
        select: { id: true },
      });
      const accessibleIdSet = new Set(accessibleDocs.map((doc) => doc.id));
      validIds = ids.filter((id) => accessibleIdSet.has(id));
    }

    const updateData = {};
    if (name !== undefined) updateData.name = name;
    if (prompt !== undefined) updateData.prompt = prompt;
    if (description !== undefined) updateData.description = description;
    if (model !== undefined) updateData.model = model;
    if (avatarUrl !== undefined) updateData.avatarUrl = avatarUrl;
    if (enabled !== undefined) updateData.enabled = Boolean(enabled);

    const updated = await prisma.$transaction(async (tx) => {
      const updatedBot = Object.keys(updateData).length > 0
        ? await tx.bot.update({
            where: { id: bot.id },
            data: updateData,
          })
        : bot;

      if (ids) {
        await tx.botDocument.deleteMany({ where: { botId: bot.id } });
        if (validIds && validIds.length > 0) {
          await tx.botDocument.createMany({
            data: validIds.map((id) => ({ botId: bot.id, documentId: id })),
          });
        }
      }

      return updatedBot;
    });

    const withDocs = await prisma.bot.findUnique({
      where: { id: updated.id },
      include: { documents: { include: { document: { select: { id: true, displayName: true } } } } },
    });

    await logEvent({
      event: "bot.updated",
      actorId: req.user.id,
      targetType: "bot",
      targetId: withDocs.id,
      meta: {
        name: withDocs.name,
        nameChanged: Boolean(name),
        promptChanged: Boolean(prompt),
        documentCount: withDocs.documents.length,
        ...getRequestContext(req),
      },
    });

    res.json({
      id: withDocs.id,
      name: withDocs.name,
      prompt: withDocs.prompt,
      description: withDocs.description,
      model: withDocs.model,
      avatarUrl: withDocs.avatarUrl,
      enabled: withDocs.enabled !== false,
      createdAt: withDocs.createdAt,
      updatedAt: withDocs.updatedAt,
      documents: withDocs.documents.map((link) => link.document),
    });
    await invalidateUserCaches(req.user.id);
  } catch (err) {
    console.error("PATCH /api/bots/:id error:", err);
    res.status(500).json({
      error: "เซิร์ฟเวอร์ตอบ 500 — backend อาจยังไม่พร้อม ลองรีเฟรชหรือตรวจสอบ Docker (api/legacy)",
      detail: err?.message || String(err),
    });
  }
});

botsRouter.delete("/:id", authenticate, async (req, res) => {
  try {
    const bot = await prisma.bot.findFirst({
      where: { id: req.params.id, ownerId: req.user.id },
    });

    if (!bot) {
      res.status(404).json({ error: "Bot not found" });
      return;
    }

    if (bot.name === HELP_BOT_NAME) {
      res.status(403).json({ error: "ไม่สามารถลบบอทช่วยสอนได้" });
      return;
    }

    await logEvent({
      event: "bot.deleted",
      actorId: req.user.id,
      targetType: "bot",
      targetId: bot.id,
      meta: { name: bot.name, ...getRequestContext(req) },
    });

    await deleteBotWithCleanup(bot.id);
    res.json({ ok: true });
    await invalidateUserCaches(req.user.id);
  } catch (error) {
    console.error("Failed to delete bot", error);
    res.status(500).json({ error: "Failed to delete bot" });
  }
});
